graph/telemetry/istio/appender/istio_details.go

```diff
@@ -164,3 +178,98 @@ func addLabels(trafficMap graph.TrafficMap, globalInfo *graph.AppenderGlobalInfo
 		}
 	}
 }
+
+func decorateGateways(trafficMap graph.TrafficMap, globalInfo *graph.AppenderGlobalInfo, namespaceInfo *graph.AppenderNamespaceInfo) {
+	// Get ingress-gateways deployments in the namespace. Then, find if the graph is showing any of them. If so, flag the GW nodes.
+	ingressWorkloads := getIngressGatewaysWorkloads(globalInfo)
+	istioAppLabelName := config.Get().IstioLabels.AppLabelName
+
+	ingressNodeMapping := make(map[*models.WorkloadListItem][]*graph.Node)
+	if ingressWorkloadsList, ok := ingressWorkloads[namespaceInfo.Namespace]; ok {
+		for _, gw := range ingressWorkloadsList {
+			for _, node := range trafficMap {
+				if node.App == gw.Labels[istioAppLabelName] && node.Namespace == namespaceInfo.Namespace {
+					node.Metadata[graph.IsIngressGw] = true
+					ingressNodeMapping[&gw] = append(ingressNodeMapping[&gw], node)
+				}
+			}
+		}
+	}
+
+	// If there is any ingress gateway node in the processing namespace, find Gateway CRDs and
+	// match them against gateways in the graph.
+	if len(ingressNodeMapping) != 0 {
+		gatewaysCrds := getIstioGatewayResources(globalInfo)
+
+		for _, gwCrd := range gatewaysCrds {
+			gwSelector := labels.Set(gwCrd.Spec.Selector).AsSelector()
+			for gw, nodes := range ingressNodeMapping {
+				if gwSelector.Matches(labels.Set(gw.Labels)) {
+
+					// If we are here, the GatewayCrd selects the Gateway workload.
+					// So, all node graphs associated with the GW workload should be listening
+					// requests for the hostnames listed in the GatewayCRD.
+
+					// Let's extract the hostnames and add them to the node metadata.
+					for _, node := range nodes {
+						gwServers := gwCrd.Spec.Servers.([]interface{})
+						var hostnames []string
+
+						for _, gwServer := range gwServers {
+							gwServerMap := gwServer.(map[string]interface{})
+							gwHosts := gwServerMap["hosts"].([]interface{})
+							for _, gwHost := range gwHosts {
+								hostnames = append(hostnames, gwHost.(string))
+							}
+						}
+
+						// Metadata format: { gatewayName => array of hostnames }
+						node.Metadata[graph.Gateways] = graph.GatewaysMetadata{
+							gwCrd.Metadata.Name: hostnames,
+						}
+					}
+				}
+			}
+		}
+	}
+}
+
+func getIngressGatewaysWorkloads(globalInfo *graph.AppenderGlobalInfo) map[string][]models.WorkloadListItem {
```

Comment:
Yes, you asked for it.
Indeed, this is no longer storing data to `globalInfo`. That parameter is only used to access `globalInfo.Business` to do queries, which isn't available in `NamespaceInfo`.

> Why not use NamespaceInfo and only query the namespaces relevant to the graph request?

Hmm... I think I can see what you are saying. Probably, this function should query only the namespace that the appender was ran for. I'll check.

Reply:
After some testing, I think we need it as is. Furthermore, I had to do a fix. See commit: https://github.com/kiali/kiali/pull/4138/commits/52ab0a76e5b627c448f5fc951c90c5d702d673ab.

The rationale is that Gateway CRDs can match workloads in any namespace (although this is configurable). See https://istio.io/latest/docs/reference/config/networking/gateway/#Gateway.

I understand that `buildTrafficMap` is ran first for certain namespace. Then the appenders are ran for that namespace. This is repeated for each NS requested in the graph. However, the `buildTrafficMap` function can return nodes that belong to other namespaces. If one of these "foreign namespace" nodes is a Gateway workload, we need to badge it and also we need to find Gateway CRDs that do a match on it. Thus, I think that `getIngressGatewayWorkloads` cannot be limited to the namespace specified in `NamespaceInfo`.

Reply:
> could pass AccessibleNamespaces as a parameter to the appender, instead of querying here.

Do you mean the `AccessibleNamespaces` in the global configuration - `config.Get().Deployment.AccessibleNamespaces`?

If you do, I don't think that's possible, because we probably want to iterate on namespaces accessible by the logged-in user (i.e. apply RBAC), not namespaces configured for Kiali.

Reply:
AccessibleNamespaces should always be relevant to the current user, I think.  We do have other appenders that take AccessibleNamespaces as a parameter, see [here](https://github.com/kiali/kiali/blob/master/graph/telemetry/istio/appender/appender.go#L61-L71), for example.  I understand that we want to badge "IsOutside" nodes and that is why we may need to query a namespace outside of the current, selected, namespace.  But, in general, we don't add additional information for namespaces inaccessible to the user. So, I'm not sure how far we want to go in that case.  Remember that you are not allowed to select an inaccessible node in the UI. Also, we don't need to badge the same outsider node twice, so we may want to check before re-running the logic for the same node on different appender runs.

Reply:
Hopefully, this is fixed in the last commit.

Reply:
>  Ultimately, I think it would be better to run this new logic after the trafficMap has been fully built. But that's not how appenders are currently working

I have considered having appenders optionally be "per-namespace" or a "finalizer".  But I thought maybe it was over-complicated.  There is actually a little bit of logic that happens after the appenders run, to decorate the final TrafficMap, but currently that's just built-into the base-graph-gen.